test/boost: speed up test test_indexing_paging_and_aggregation by making internal page size configurable

nyh opened the pull request:
The C++ test `test_indexing_paging_and_aggregation` is one of the slowest tests in test/boost. The reason for its slowness is that it needs a table with more rows than SELECT's "DEFAULT_COUNT_PAGE_SIZE" which was hard-coded to 10,000, so the test needed to write and read tens of thousands of rows, and did it multiple times.

It turns out the code actually had an ad-hoc mechanism to override DEFAULT_COUNT_PAGE_SIZE in a C++ test, but both this mechanism and the test itself were so opaque I didn't find it until I fixed it in a different way: What I ended up doing in this pull request is the following (each step in a separate patch):

1. Rewrite this test in Python, in the test/cqlpy framework. This was straightforward, as this test only used CQL and not internal interfaces. The reason why this test wasn't written in Python in the first place is that it was written in 2019, a year before cqlpy existed. A added extensive comments to the new tests, and I finally understood what it was doing :-)
2. I replaced the ad-hoc C++-test-only mechanism of overriding DEFAULT_COUNT_PAGE_SIZE by a bona-fide configuration parameter, `select_internal_page_size`.
3. Finally, the Python test can temporarily lower `select_internal_page_size` and use a table with much fewer rows.

After this series, the test `test_indexing_paging_and_aggregation` (which is now in Python instead of C++) takes around half a second, 20 times faster than before. I expect the speedup to be even more dramatic for the debug build.

nyh commented on test/cqlpy/test_secondary_index.py:
```diff
@@ -2100,3 +2100,69 @@ def test_limit_partition_slice(cql, test_keyspace):
+        # select_internal_page_size rows at a time, but even though
+        # we have more rows than that in the table, it must only return a
+        # single result row when all the data was aggregated - the CQL API
+        # doesn't allow it to return return partial, per-page, results.
```
In case Copilot or another reader is not familiar with the Gauss story, here's a link: https://letstalkscience.ca/educational-resources/backgrounders/gauss-summation

nyh replied:
Added a whole proof in the comment, plus a remark that the margin is too narrow for the proofs of the other forumulas :-)

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -431,7 +428,7 @@ select_statement::do_execute(query_processor& qp,
     if (aggregate || nonpaged_filtering) {
-        page_size = page_size <= 0 ? internal_paging_size : page_size;
+        page_size = page_size <= 0 ? qp.db().get_config().select_internal_page_size() : page_size;
     }
```
Why is anything needed? If you supply page_size > 0 (which should be common) isn't it enough?

nyh replied:
I'm not sure what you're asking - it seems this code does what you suggest: If page_size is supplied by the user, it is used for the aggregation. Just if page_size=0, we use this default "internal page size".

However, I should note that this logic was written 7 years ago by Piotr Sarna, and I'm not sure I fully understand it or in which queries it actually applies or why. In a bunch of other places in select_statement.cc, select_internal_page_size is used unconditionally for internal reads.

In this patch, I did not try to fix or "improve" anything in the scanning logic - just to make the constant defaulting to 10,000 be configurable, so tests for it can be shorter.

avikivity replied:
@nyh please see this

nyh replied:
First of all, I don't know the full answer - as I said this patch just makes a constant configurable, and I did not try to second-guess all existing uses of this constant.

I'm not sure if when you ask "why is this constant **ever** used", the "ever" refers to the specific line here and the `page_size<=0` condition, or "ever" in the sense in any place. As I noted, in *other* places, this constant is used because it seems to be used unconditionally, regardless of what the driver passed. So I'll assume you are asking just on this specifc line you highlighted:

You're right that in this specific line it seems that if the driver pass a page size, then indeed it will be used and not the configuration. So is this "?" stuff here even needed? Probably not in practice. But we **do need to pick** a page size in this code if the driver passed a 0 page size (or maybe an internal query with no paging)! We can't use "0" in that case, right? So the default "internal page size" makes sense to use, doesn't it?

avikivity commented on cql3/statements/select_statement.hh:
```diff
@@ -243,7 +243,8 @@ private:
-            lw_shared_ptr<const service::pager::paging_state> paging_state) const;
+            lw_shared_ptr<const service::pager::paging_state> paging_state,
+            uint32_t internal_page_size) const;
 
```
Let's remove DEFAULT_COUNT_PAGE_SIZE from .hh, since it's no longer used.

nyh replied:
Yes, sorry, I missed it.

temichus reviewed (commented):
This PR does not touch test.py core, only the test body level. nothing to review for me here

nyh commented:
> This PR does not touch test.py core, only the test body level. nothing to review for me here

I asked you to "review" this only in the sense that this is a slow test that you complained about, so I thought you'd want to know I made it faster - and perhaps how.

nyh commented:
Yet again, I have no idea why the CI failed - I don't see any specific test failing, and can't find errors in a log file with myriads of lines :-( I'll just restart the CI...

piodul commented:
> #### Failed Tests (1/79160):
> 
>     * [test_one_big_mutation_corrupted_on_startup.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19189/testReport/junit/cluster.dtest.commitlog_test/TestCommitLog/dev___test_py___dev___test_one_big_mutation_corrupted_on_startup_dev_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_one_big_mutation_corrupted_on_startup.dev.1)

Looks like https://github.com/scylladb/scylladb/issues/25627 - known flaky test

piodul commented on test/boost/secondary_index_test.cc:
```diff
@@ -1416,7 +1415,7 @@ SEASTAR_TEST_CASE(test_secondary_index_on_non_pk_ck_column_and_aggregation) {
 
     return do_with_cql_env_thread([] (cql_test_env& e) {
-        cql3::statements::set_internal_paging_size_guard g(page_scenarios_page_size);
+        e.db_config().select_internal_page_size.set(page_scenarios_page_size);
```
We probably need to set the option on all shards, not only the local one, right?

nyh commented:
Rebased to resolve conflicts, and also fixed @piodul's review comment - start the cql environment with a different configuration (a trick I copied from another test that does this) instead trying to change the configuration after the environment is already up on who many knows how many threads.
